R/error-checks.R

```diff
@@ -285,46 +285,60 @@ logErrorMessage <- function(message, logFolderPath = getwd()) {
+#' Units for `dv `and `time` need to be defined at least once in either
+#' the observed dataset, its dictionary or outputs
+#' In case of multiple definitions, warnings will thrown and the following priorities will be applied:
+#' 1. Use units from outputs
+#' 2. Use units from observed dataset
+#' 3. Use units from dictionary
+#' @param observedMetaDataFile Path of meta data file on observed dataset (also called dictionary)
+#' @param observedDataFile Path of observed dataset
+#' @param outputs list or array of `Output` objects
+validateObservedMetaDataFile <- function(observedMetaDataFile, observedDataFile, outputs) {
   # Check that dictionary is provided
   if (isOfLength(observedMetaDataFile, 0)) {
     stop(messages$errorObservedMetaDataFileNotProvided(observedDataFile))
   }
   # Read dictionary and check that mandatory variables are included
   dictionary <- readObservedDataFile(observedMetaDataFile)
-  if (!isIncluded(dictionaryParameters$nonmemUnit, names(dictionary))) {
-    dictionary[, dictionaryParameters$nonmemUnit] <- NA
+  if (!isIncluded(dictionaryParameters$datasetUnit, names(dictionary))) {
+    dictionary[, dictionaryParameters$datasetUnit] <- NA
   }
-  validateIsIncluded(c(dictionaryParameters$ID, dictionaryParameters$nonmenColumn), names(dictionary))
-  validateIsIncluded(c(dictionaryParameters$timeID, dictionaryParameters$dvID), dictionary[, dictionaryParameters$ID])
+  validateIsIncludedInDataset(c(dictionaryParameters$ID, dictionaryParameters$datasetColumn), dictionary, datasetName = "dictionary")
+  validateIsIncluded(c(dictionaryParameters$timeID, dictionaryParameters$dvID), dictionary[, dictionaryParameters$ID], groupName = paste0("Column '", dictionaryParameters$ID, "'"))
 
   # Check that dictionary and observed data are consitent
   observedDataset <- readObservedDataFile(observedDataFile)
   timeVariable <- getDictionaryVariable(dictionary, dictionaryParameters$timeID)
   dvVariable <- getDictionaryVariable(dictionary, dictionaryParameters$dvID)
   lloqVariable <- getDictionaryVariable(dictionary, dictionaryParameters$lloqID)
 
-  checkIsIncluded(c(timeVariable, dvVariable), names(observedDataset))
-  checkIsIncluded(lloqVariable, names(observedDataset), nullAllowed = TRUE)
+  checkIsIncludedInDataset(c(timeVariable, dvVariable), observedDataset, datasetName = "observed dataset")
+  checkIsIncludedInDataset(lloqVariable, observedDataset, datasetName = "observed dataset", nullAllowed = TRUE)
 
-  # Units
-  # If unit is defined as a value in nonmemUnit
-  timeMapping <- dictionary[, dictionaryParameters$ID] %in% dictionaryParameters$timeID
-  dvMapping <- dictionary[, dictionaryParameters$ID] %in% dictionaryParameters$dvID
-
-  timeUnit <- as.character(dictionary[timeMapping, dictionaryParameters$nonmemUnit])
-  dvUnit <- as.character(dictionary[dvMapping, dictionaryParameters$nonmemUnit])
+  # Check of unit definitions:
+  # 1) unit defined in outptuts
+  dataUnit <- NULL

```

Comment:
isn't a list? E.g. `dataUnits`?

Reply:
From what I have understood, 3 different ways can be used to define the units of an observed dataset:
- Definition of units in dictionary file by filling the columns **datasetUnit** for **time** and **dv**
- Definition of units in dictionary by filling the column **datasetColumn** for **timeUnit** and **dvUnit** (this definition maps an observed data column for units and allow multiple units for a same dataset)
- Definition of **dv** units for the selected data of `Output` objects using the new field `dataUnit` (this definition maps subsets of observed data to their units and allow multiple units for a same dataset)

Since one definition is necessary, I need to throw an error in case of no definition (with a message reminding these 3 methods) and throw a warning in case of multiple definitions (reminding which methods get priority).

Since **datasetUnit** and **datasetColumn** are either a string value or `NA`, `NULL`, I wanted a similar format for the units defined by `Output` objects. So, `dataUnit` is first `NULL` , and if `Output` objects define `dataUnit`, I get an array of characters using `unlist(lapply(...))` (`sapply` can't be used here in case one of the `Output` objects did not define `dataUnit`, I'll get a list and not the character array).
Then, I can concatenate these definitions as an array of characters and check for its length.

Reply:
ok cool